README.md
# WhatWaf

## 概述
### Overview
WhatWaf 是一个高级的防火墙检测工具，用于检测当前Web url 或者 Web应用使用了什么Waf，通过特定的选项可以使用特定的 payload 绕过防火墙，剩下的靠自己去挖掘了

## 免责声明
### Disclaimer
此项目来源为 [https://github.com/Ekultek/WhatWaf](https://github.com/Ekultek/WhatWaf)

Origin sourc: [https://github.com/Ekultek/WhatWaf](https://github.com/Ekultek/WhatWaf)

出于学习的目的，在遵循源项目的开源协议的基础上修改整体架构，使其更加符合软件工程的规范，并且添加很多必要的注释，修复一些bug，原功能基本不变，只做一些更改, 更改的条目在 CHANGELOG 文件中

## 所有的选项以及说明
### Options
```
                              ,------.
                             '  .--.  '
    ,--.   .--.   ,--.   .--.|  |  |  |
    |  |   |  |   |  |   |  |'--'  |  |
    |  |   |  |   |  |   |  |    __.  |
    |  |.'.|  |   |  |.'.|  |   |   .'
    |         |   |         |   |___|
    |   ,'.   |hat|   ,'.   |af .---.
    '--'   '--'   '--'   '--'   '---'
"/><script>alert("WhatWaf?<|>v1.4.5($dev)");</script>

usage: ./whatwaf.py [options] [arguments]

optional arguments:
  -h, --help            show this help message and exit

mandatory arguments:
  arguments that have to be passed for the program to run

  -u URL, --url URL     Pass a single URL to detect the protection
  -l PATH, --list PATH, -f PATH, --file PATH
                        Pass a file containing URL's (one per line) to detect
                        the protection
  -b FILE-PATH, --burp FILE-PATH
                        Pass a Burp Suite request file to perform WAF
                        evaluation

request arguments:
  arguments that will control your requests

  --pa USER-AGENT       Provide your own personal agent to use it for the HTTP
                        requests
  --ra                  Use a random user-agent for the HTTP requests
  -H HEADER=VALUE,HEADER:VALUE.., --headers HEADER=VALUE,HEADER:VALUE..
                        Add your own custom headers to the request. To use
                        multiple separate headers by comma. Your headers need
                        to be exact(IE: Set-Cookie=a345ddsswe,X-Forwarded-
                        For:127.0.0.1)
  --proxy PROXY         Provide a proxy to run behind in the format
                        type://address:port (eg socks5://10.54.127.4:1080
  --tor                 Use Tor as the proxy to run behind, must have Tor
                        installed
  -p PAYLOADS, --payloads PAYLOADS
                        Provide your own payloads separated by a comma IE AND
                        1=1,AND 2=2
  --pl PAYLOAD-LIST-PATH
                        Provide a file containing a list of payloads 1 per
                        line
  --force-ssl           Force the assignment of HTTPS instead of HTTP while
                        processing (*default=HTTP unless otherwise specified
                        by URL)
  --throttle THROTTLE-TIME (seconds)
                        Provide a sleep time per request (*default=0)
  --timeout TIMEOUT     Control the timeout time of the requests (*default=15)
  -P, --post            Send a POST request (*default=GET)
  -D POST-STRING, --data POST-STRING
                        Send this data with the POST request (IE
                        password=123&name=Josh *default=random)
  -t threaded, --threaded threaded
                        Send requests in parallel (specify number of threads
                        *default=1)
  -tP CONFIGTORPORT, --tor-port CONFIGTORPORT
                        Change the port that Tor runs on (*default=9050)

encoding options:
  arguments that control the encoding of payloads

  -e PAYLOAD [TAMPER-SCRIPT-LOAD-PATH ...], --encode PAYLOAD [TAMPER-SCRIPT-LOAD-PATH ...]
                        Encode a provided payload using provided tamper
                        script(s) you are able to payy multiple tamper script
                        load paths to this argument and the payload will be
                        tampered as requested
  -el PATH TAMPER-SCRIPT-LOAD-PATH, --encode-list PATH TAMPER-SCRIPT-LOAD-PATH
                        Encode a file containing payloads (one per line) by
                        passing the path and load path, data can only encoded
                        using a single tamper script load path

output options:
  arguments that control how WhatWaf handles output

  -F, --format          Format the output into a dict and display it
  -J, --json            Send the output to a JSON file
  -Y, --yaml            Send the output to a YAML file
  -C, --csv             Send the output to a CSV file
  --fingerprint         Save all fingerprints for further investigation
  --tamper-int INT      Control the amount of tampers that are displayed
                        (*default=5)
  --traffic FILENAME    store all HTTP traffic headers into a file of your
                        choice
  --force-file          Force the creation of a file even if there is no
                        protection identified
  -o DIR, --output DIR  Save a copy of the file to an arbitrary directory

database arguments:
  arguments that pertain to Whatwafs database

  -uC, --view-url-cache
                        Display all the URL cache inside of the database, this
                        includes the netlock, tamper scipts, webserver, and
                        identified protections
  -pC, --payload-cache  View all payloads that have been cached inside of the
                        database
  -vC, --view-cache     View all the cache in the database, everything from
                        URLs to payloads
  --export FILE-TYPE    Export the already encoded payloads to a specified
                        file type and save them under the home(~/.whatwaf)
                        directory

misc arguments:
  arguments that don't fit in any other category

  --verbose             Run in verbose mode (more output)
  --update              Update WhatWaf to the newest development version
  --save FILENAME       Save the encoded payloads into a file
  --skip                Skip checking for bypasses and just identify the
                        firewall
  --verify-num INT      Change the request amount to verify if there really is
                        not a WAF present(*default=5)
  -W, --determine-webserver
                        Attempt to determine what web server is running on the
                        backend (IE Apache, Nginx, etc.. *default=False)
  --wafs                Output a list of possible firewalls that can be
                        detected by this program
  --tampers             Output a list of usable tamper script load paths

  --clean               Clean up WhatWaf home folders
```

## 下载和安装
### Installation
```
git clone https://github.com/Lunpopo/WhatWaf.git

cd WhatWaf 

pip install -r requirements.txt

python whatwaf.py -h
```

## 使用
### Usage

#### Demon Video: [![to_video](http://i67.tinypic.com/2daawow.png)](https://vimeo.com/247623511)

#### Proof of concept
First we'll run the website through WhatWaf and figure out which firewall protects it (if any):
![item1](http://i67.tinypic.com/142y9s6.png)

Next we'll go to that website and see what the page looks like:
![item2](http://i64.tinypic.com/262mjhl.png)

Hmm.. that doesn't really look like Cloudflare does it? Lets see what the headers say:
![item4](http://i66.tinypic.com/5txx5x.png)

And finally, lets try one of the bypasses that it tells us to try:
![item3](http://i66.tinypic.com/sdi3x0.png)



CHANGELOG
## 这是与原版的差异记录文件

1. 修改了部分架构，另其更加符合软件工程规范
2. omit --hide options 因为根本就没什么卵用
3. 从 settings 中删除 DEFAULT_USER_AGENT 
4. 从 command line 中删除 --check-tor 选项，默认带上它
5. 从 command line 中删除 -c --url-cache, 默认带上它，没必要特意指定
6. 从 command line 中删除 -T --test default = True, 默认带上它，没必要特意指定
7. 从 command line 中删除 --fingerprint 没什么卵用


lib/cmdline.py
# encoding: utf8
from argparse import ArgumentParser, Action


class StoreDictKeyPairs(Action):
    """
    custom action to create a dict from a provided string in the format of key=value or key:value
    """

    retval = {}

    def __call__(self, parser, namespace, values, option_string=None):
        # discover what we split by
        """
        保存 cmd 中 dic 参数
        :param parser:
        :param namespace:
        :param values:
        :param option_string:
        :return:
        """
        for kv in values.split(","):
            if ":" in kv:
                splitter = ":"
            else:
                splitter = "="
            if kv.count(splitter) != 1:
                first_equal_index = kv.index(splitter)
                key = kv[:first_equal_index].strip()
                value = kv[first_equal_index + 1:].strip()
                self.retval[key] = value
            else:
                k, v = kv.split(splitter)
                self.retval[k.strip()] = v.strip()
        # return the attribute as {'foo': 'bar'}
        setattr(namespace, self.dest, self.retval)


class CmdLineParser(ArgumentParser):
    def __init__(self):
        super(CmdLineParser, self).__init__()

    @staticmethod
    def cmd_parser(get_help=False):
        """
        # prog 程序的名字，默认 sys.argv[0]
        # usage -> 描述程序的用法，默认是展示options的描述

        :return: 返回 argparser 的 opts object
        """

        # apparently Python 3.x doesn't like it when there's a '%' in the string
        # this will cause some issues, more specifically a `TypeError` because it's trying
        # to format the string. so the simplest solution for the problem? remove the '%'
        # sign from the string. because Python 3.x likes to be as difficult and ridiculous
        # as it possibly can.
        parser = ArgumentParser(prog="whatwaf.py", add_help=True, usage=(
            "./whatwaf.py [options] [arguments]"
        ))

        # helper argument
        # helper = parser.add_argument_group('optional arguments')
        # helper.add_argument('-h', '--help', desc='show this help message and exit', action='help')

        # mandatory argument
        mandatory = parser.add_argument_group("mandatory arguments",
                                              "arguments that have to be passed for the program to run")
        mandatory.add_argument("-u", "--url", dest="runSingleWebsite", metavar="URL",
                               help="Pass a single URL to detect the protection")
        mandatory.add_argument("-l", "--list", "-f", "--file", dest="runMultipleWebsites",
                               metavar="PATH", default=None,
                               help="Pass a file containing URL's (one per line) to detect the protection")
        mandatory.add_argument("-b", "--burp", dest="burpRequestFile", metavar="FILE-PATH", default=None,
                               help="Pass a Burp Suite request file to perform WAF evaluation")

        # request argument
        req_args = parser.add_argument_group("request arguments",
                                             "arguments that will control your requests")
        req_args.add_argument("--pa", dest="usePersonalAgent", metavar="USER-AGENT",
                              help="Provide your own personal agent to use it for the HTTP requests")
        req_args.add_argument("--ra", dest="useRandomAgent", action="store_true",
                              help="Use a random user-agent for the HTTP requests")
        req_args.add_argument("-H", "--headers", dest="extraHeaders", action=StoreDictKeyPairs,
                              metavar="HEADER=VALUE,HEADER:VALUE..",
                              help="Add your own custom headers to the request. To use multiple "
                                   "separate headers by comma. Your headers need to be exact"
                                   "(IE: Set-Cookie=a345ddsswe,X-Forwarded-For:127.0.0.1)")
        req_args.add_argument("--proxy", dest="runBehindProxy", metavar="PROXY",
                              help="Provide a proxy to run behind in the format "
                                   "type://address:port (eg socks5://10.54.127.4:1080")
        req_args.add_argument("--tor", dest="runBehindTor", action="store_true",
                              help="Use Tor as the proxy to run behind, must have Tor installed")
        # 默认带上, 没必要特意带上
        # req_args.add_argument("--check-tor", dest="checkTorConnection", action="store_true",
        #                       help="Check your Tor connection")
        req_args.add_argument("-p", "--payloads", dest="providedPayloads", metavar="PAYLOADS",
                              help="Provide your own payloads separated by a comma IE AND 1=1,AND 2=2")
        req_args.add_argument("--pl", dest="payloadList", metavar="PAYLOAD-LIST-PATH",
                              help="Provide a file containing a list of payloads 1 per line")
        req_args.add_argument("--force-ssl", dest="forceSSL", action="store_true",
                              help="Force the assignment of HTTPS instead of HTTP while processing "
                                   "(*default=HTTP unless otherwise specified by URL)")
        req_args.add_argument("--throttle", dest="sleepTimeThrottle", type=int, metavar="THROTTLE-TIME (seconds)",
                              default=0, help="Provide a sleep time per request (*default=0)")
        req_args.add_argument("--timeout", dest="requestTimeout", type=int, metavar="TIMEOUT", default=15,
                              help="Control the timeout time of the requests (*default=15)")
        req_args.add_argument("-P", "--post", dest="postRequest", action="store_true",
                              help="Send a POST request (*default=GET)")
        req_args.add_argument("-D", "--data", dest="postRequestData", metavar="POST-STRING",
                              help="Send this data with the POST request "
                                   "(IE password=123&name=Josh *default=random)")
        req_args.add_argument("-t", "--threaded", dest="threaded",  metavar="threaded",  type=int,
                              help="Send requests in parallel (specify number of threads *default=1)")
        req_args.add_argument("-tP", "--tor-port", type=int, default=9050, dest="configTorPort",
                              help="Change the port that Tor runs on (*default=9050)")
        # 默认带上, 没必要特意带上
        # req_args.add_argument("-T", "--test", dest="testTargetConnection", default=True, action="store_false",
        #                       help="Test the connection to the website before starting (default is True)")

        # encoding argument 组
        encoding_opts = parser.add_argument_group("encoding options",
                                                  "arguments that control the encoding of payloads")
        encoding_opts.add_argument("-e", "--encode",
                                   dest="encodePayload", nargs="+", metavar=("PAYLOAD", "TAMPER-SCRIPT-LOAD-PATH"),
                                   help="Encode a provided payload using provided tamper script(s) "
                                        "you are able to payy multiple tamper script load paths to "
                                        "this argument and the payload will be tampered as requested")
        encoding_opts.add_argument("-el", "--encode-list",
                                   dest="encodePayloadList",
                                   nargs=2,
                                   metavar=("PATH", "TAMPER-SCRIPT-LOAD-PATH"),
                                   help="Encode a file containing payloads (one per line) "
                                        "by passing the path and load path, data can only "
                                        "encoded using a single tamper script load path")

        # output argument 组
        output_opts = parser.add_argument_group("output options",
                                                "arguments that control how WhatWaf handles output")
        output_opts.add_argument("-F", "--format", action="store_true", dest="formatOutput",
                                 help="Format the output into a dict and display it")
        output_opts.add_argument("-J", "--json", action="store_true", dest="sendToJSON",
                                 help="Send the output to a JSON file")
        output_opts.add_argument("-Y", "--yaml", action="store_true", dest="sendToYAML",
                                 help="Send the output to a YAML file")
        output_opts.add_argument("-C", "--csv", action="store_true", dest="sendToCSV",
                                 help="Send the output to a CSV file")
        output_opts.add_argument("--fingerprint", action="store_true", dest="saveFingerprints",
                                 help="Save all fingerprints for further investigation")
        output_opts.add_argument("--tamper-int", metavar="INT", dest="amountOfTampersToDisplay", type=int,
                                 default=5, help="Control the amount of tampers that are displayed (*default=5)")
        output_opts.add_argument("--traffic", metavar="FILENAME", dest="trafficFile",
                                 help="store all HTTP traffic headers into a file of your choice")
        output_opts.add_argument("--force-file", action="store_true", default=False, dest="forceFileCreation",
                                 help="Force the creation of a file even if there is no protection identified")
        output_opts.add_argument("-o", "--output", metavar="DIR", dest="outputDirectory", default=None,
                                 help="Save a copy of the file to an arbitrary directory")

        # database argument 组
        database_arguments = parser.add_argument_group("database arguments",
                                                       "arguments that pertain to Whatwafs database")
        # 默认带上, 没必要特意指定
        # database_arguments.add_argument(
        #     "-c", "--url-cache", default=False, action="store_true", dest="checkCachedUrls",
        #     help="Check against URL's that have already been cached into the database before running them "
        #          "saves some time on scanning multiple (*default=False)"
        # )
        database_arguments.add_argument(
            "-uC", "--view-url-cache", default=False, action="store_true", dest="viewUrlCache",
            help="Display all the URL cache inside of the database, this includes the netlock, "
                 "tamper scipts, webserver, and identified protections"
        )
        database_arguments.add_argument(
            "-pC", "--payload-cache", action="store_true", default=False, dest="viewCachedPayloads",
            help="View all payloads that have been cached inside of the database"
        )
        database_arguments.add_argument(
            "-vC", "--view-cache", action="store_true", default=False, dest="viewAllCache",
            help="View all the cache in the database, everything from URLs to payloads"
        )
        database_arguments.add_argument(
            "--export", metavar="FILE-TYPE", default=None, dest="exportEncodedToFile",
            choices=["txt", "text", "json", "csv", "yaml", "yml"],
            help="Export the already encoded payloads to a specified file type and save them "
                 "under the home(~/.whatwaf) directory"
        )

        # misc argument 组
        misc = parser.add_argument_group("misc arguments",
                                         "arguments that don't fit in any other category")
        misc.add_argument("--verbose", dest="runInVerbose", action="store_true",
                          help="Run in verbose mode (more output)")
        misc.add_argument("--update", dest="updateWhatWaf", action="store_true",
                          help="Update WhatWaf to the newest development version")
        misc.add_argument("--save", dest="saveEncodedPayloads", metavar="FILENAME",
                          help="Save the encoded payloads into a file")
        misc.add_argument("--skip", dest="skipBypassChecks", action="store_true",
                          help="Skip checking for bypasses and just identify the firewall")
        misc.add_argument("--verify-num", dest="verifyNumber", metavar="INT", type=int,
                          help="Change the request amount to verify if there really is not a WAF present"
                               "(*default=5)")
        misc.add_argument("-W", "--determine-webserver", action="store_true", default=False, dest="determineWebServer",
                          help="Attempt to determine what web server is running on the backend "
                               "(IE Apache, Nginx, etc.. *default=False)")
        misc.add_argument("--wafs", action="store_true", default=False, dest="viewPossibleWafs",
                          help="Output a list of possible firewalls that can be detected by this program")
        misc.add_argument("--tampers", action="store_true", dest="listEncodingTechniques",
                          help="Output a list of usable tamper script load paths")

        # hidden argument
        hidden = parser.add_argument_group()
        hidden.add_argument("--clean", action="store_true", dest="cleanHomeFolder",
                            help='Clean up WhatWaf home folders')

        opts = parser.parse_args()

        if get_help:
            return parser.print_help()
        else:
            return opts

    @staticmethod
    def parse_help_menu(data, start_field, end_field):
        """
        parse the help menu from a certain string to a certain string
        and return the parsed help
        将 help 中的文字截取出来

        :param data: cmd 中所有的 help 信息
        :param start_field: 以...子模块开头的块, 例如 -> encoding options:
        :param end_field: 以...子模块结束的块, 例如-> output options:
        :return:
        """

        try:
            start_index = data.index(start_field)
            end_index = data.index(end_field)
            retval = data[start_index:end_index].strip()
        except TypeError:
            # python3 is stupid and likes `bytes` because why tf not?
            plus = 60
            # so now we gotta dd 60 in order to get the last line from the last command
            # out of the way
            start_index = data.decode().index(start_field) + plus
            end_index = data.decode().index(end_field)
            # and then we gotta convert back
            data = str(data)
            # and then we gotta store into a temporary list
            tmp = data[start_index:end_index]
            # split the list into another list because of escapes
            # join that list with a new line and finally get the
            # retval out of it. Because that makes PERFECT sense
            retval = "\n".join(tmp.split("\\n"))
        return retval

